chore: Update Key Vault documentation with latest changes (#345)
CluedIn-io/CluedIn.Documentation: docs/kb/kb0004-akv2aks.md

    When we get later into the guide, we'll explain how to pair these back up into a single object.

> 1. With all your desired secrets and certificates now uploaded to AKV, it's time to configure the `Values.yaml` to begin synchronising.

Reviewer: ```suggestion
1. With all your desired secrets and certificates now uploaded to AKV, it's time to configure the `Values.yaml` to begin synchronizing.
```

        userAssignedIdentityID: $kvManagedIdentity # This is the guid for the kv managed identity
        keyvaultName: $kv.name # This is your key vault name
        tenantId: $tenantId # This is the guid of your tenant
        secretProviderClasses:

>           cluedin-server: # For every key under `secretProviderClasses` a new secret provider class will be created with the same name an `-sync` appended.

Reviewer: ```suggestion
          cluedin-server: # For every key under `secretProviderClasses`, a new secret provider class will be created with the same name and `-sync` appended.
```

>           - secretName: cluedin-login-details # This is how the secret will appear within Kubernetes Secrets once synchronised

Reviewer: ```suggestion
          - secretName: cluedin-login-details # This is how the secret will appear within Kubernetes Secrets once synchronized.
```

>             secretType: # [OPTIONAL] Defaults to Opaque. But you can specify any supported secretType

Reviewer: ```suggestion
            secretType: # [OPTIONAL] Defaults to Opaque. But you can specify any supported secretType.
```

            secretKeys:
              password: cluedin-login-password # The left side (Key) is the name in the Kubernetes secret object. The right side (Value) is the AKV reference which will grab its value.
              username: cluedin-login-username

    infrastructure:
      cert-manager:
        #enabled: false # Only set to false if using an uploaded frontend certificate.
    ```

>     **NOTE**: If you are using the certificate upload as part of your setup, you **must** disable cert-manager by setting `enabled: false`. You also must set `frontendCrt` to a value. It will be mounted as `cluedin-frontend-crt` on `cluedin-server`. This secret name cannot change.

Reviewer: ```suggestion
    {:.important}
    If you are using the certificate upload as part of your setup, you **must** disable cert-manager by setting `enabled: false`. You also must set `frontendCrt` to a value. It will be mounted as `cluedin-frontend-crt` on `cluedin-server`. This secret name cannot change.
```

> 1. With all the secrets and certificates now done, the last step is to mount, map, and synchronise the secrets to kubernetes.

Reviewer: ```suggestion
1. With all the secrets and certificates now done, the last step is to mount, map, and synchronize the secrets to Kubernetes.
```

    For every key under `secretProviderClass`, a new secret provider will be created with the same name with `-sync` appended. 

>     This gives you the flexability to have secrets to share the same lifecycle as an application, or to use a singular pod to synchronise the secrets.

Reviewer: ```suggestion
    This gives you the flexibility to have secrets to share the same life cycle as an application, or to use a singular pod to synchronize the secrets.
```

    `cluedin-server` is the only server that will automount if keys are specified under the `cluedin-server` secretProviderClasses section.
    For everything else, a specific mount point will be required along with preventing local password creation by using the appropriate key.

    We **highly recommend** having secrets share life cycles with the application. We will cover only this scenario below.

1. Please see the mapping table below:

>     | **Application** | **Secret** | **Helm Path** | Mount Point |

Reviewer: ```suggestion
    | Application | Secret | Helm Path | Mount Point |
```

    | cluedin-elasticsearch | elasticsearch-credentials | `infrastructure.elasticsearch.auth.existingSecret` | `infrastructure.elasticsearch.extraVolumes`<br/>`infrastructure.elasticsearch.extraVolumeMounts` |
    | cluedin-redis | cluedin-redis | `infrastructure.redis.auth.existingSecret` | `infrastructure.redis.master.extraVolumes`<br/>`infrastructure.redis.master.extraVolumeMounts` |
    | cluedin-neo4j | cluedin-neo4j-secrets | `infrastructure.neo4j.neo4j.passwordFromSecret` | `infrastructure.neo4j.additionalVolumes`<br/>`infrastructure.neo4j.additionalVolumeMounts` |
    | | cluedin-neo4j-auth | `!! shared from above !!` | |
    | cluedin-grafana | cluedin-grafana | `infrastructure.monitoring.grafana.admin.existingSecret` | `infrastructure.monitoring.grafana.extraContainerVolumes`<br/>`infrastructure.monitoring.grafana.sidecar.dashboards.extraMounts` |
    | cluedin-rabbitmq | cluedin-rabbitmq | `infrastructure.rabbitmq.auth.existingPasswordSecret` | `infrastructure.rabbitmq.extraVolumes`<br/>`infrastructure.rabbitmq.extraVolumeMounts` | 
    | | cluedin-rabbitmq-load-definition | `!! shared from above !!` | |

> Please also find a sample yaml file <a href="../../assets/other/akv-sync-sample.yaml" download>here</a> which can be used as a reference for a complete setup

Reviewer: ```suggestion
Please find a sample yaml file <a href="../../assets/other/akv-sync-sample.yaml" download>here</a>, which can be used as a reference for a complete setup.
```

    Please ensure this is checked before proceeding as it may cause the cluster to not function correctly. 

- The way the secrets are synchronized is by using the `cluedin-server` pod mounting the secret store or mapping and mounting on other pods. The sample file and table above explain how to achieve this. It's important to note that the secrets synchronize only when the pod is active. It doesn't need to be in a running state, but it must at least be pending.  

    This is a limitation of the CSI driver itself rather than the solution provided by CluedIn.

- The secrets synchronized do not overwrite existing secrets that are created by the CluedIn Installer. If your secret matches the same name (front-end certificate is mandatory here), you must remove the existing secret for the synchronized secret to take over. 

- Please ensure that you have enough resources. For example, by default the Neo4j and Elasticsearch pods consume a majority of the nodes they have been assigned. Having the additional Key Vault pods on these nodes may potentially prevent these from starting up even though the Key Vault pods request very little resource.

> # Known Issues

Reviewer: ```suggestion
# Known issues

```

This section will explain some of the known issues.

- **Problem**: When doing a migration of RabbitMQ from local kubernetes password to a synced password, you must do a sequence of steps due to the RabbitMQ charts logic.

>   **Solution**:

Reviewer: ```suggestion
  **Solution**:
  
```

  1. Do an initial deployment where the secret is mounted and mapped to RabbitMQ pod as well as the initial password being supplied in the User Supply Values for RabbitMQ. 

  1. Once deployed and you can see the secretProviderClass, delete the existing rabbitMQ secrets and then kill the RabbitMQ pod. A new pod should spawn and the secrets should then be mapped to the secretProviderClass.

>   1. When this has happened, update User Supply Values to remove the password and update it to use an existing one. Redeploy the config, and this time it should succeed as the sychronised secret will exist at deployment time.

Reviewer: ```suggestion
  1. Update User Supply Values to remove the password and update it to use an existing one. Redeploy the config, and this time it should succeed as the sychronised secret will exist at deployment time.
```

  **Note**: This is only an issue with migration. Fresh installs do not have this issue.

- **Problem**: Not all secrets work from a synced source.

>   **Solution**: Unfortunately not all secrets will work from AKV. An example of this is the `acr-registry-credentials` secret. For a secret to be synchronised, it must first be mounted to a pod. However, you cannot pull the image for the pod if the secret does not exist.

Reviewer: ```suggestion
  **Solution**: Unfortunately, not all secrets will work from AKV. An example of this is the `acr-registry-credentials` secret. For a secret to be synchronized, it must first be mounted to a pod. However, you cannot pull the image for the pod if the secret does not exist.
```
